crates/http/src/body.rs

```diff
@@ -1,149 +1,84 @@
+    Quic(crate::backend::h3::body::QuicIncomingBody<h3_quinn::BidiStream<Bytes>>),
 }
 
-impl std::fmt::Debug for IncomingBody {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("IncomingBody").finish()
+#[cfg(any(feature = "http1", feature = "http2"))]
+#[cfg_attr(docsrs, doc(cfg(any(feature = "http1", feature = "http2"))))]
+impl From<hyper::body::Incoming> for IncomingBody {
+    fn from(body: hyper::body::Incoming) -> Self {
+        IncomingBody::Hyper(body)
     }
 }
 
-#[derive(derive_more::From)]
-pub(crate) enum IncomingBodyInner {
-    #[cfg(feature = "hyper")]
-    Tcp(#[from] hyper::body::Incoming),
-    #[cfg(feature = "h3")]
-    Quic(#[from] QuicIncomingBody),
-    #[cfg_attr(not(any(feature = "http3", feature = "http1", feature = "http2")), allow(dead_code))]
-    Empty,
+#[cfg(feature = "http3")]
+#[cfg_attr(docsrs, doc(cfg(feature = "http3")))]
+impl From<crate::backend::h3::body::QuicIncomingBody<h3_quinn::BidiStream<Bytes>>> for IncomingBody {
+    fn from(body: crate::backend::h3::body::QuicIncomingBody<h3_quinn::BidiStream<Bytes>>) -> Self {
+        IncomingBody::Quic(body)
+    }
 }
 
 impl http_body::Body for IncomingBody {
     type Data = Bytes;
-    type Error = crate::Error;
-
-    fn poll_frame(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
-        #[cfg(not(any(feature = "http3", feature = "http1", feature = "http2")))]
-        let _ = cx;
-
-        match &mut self.inner {
-            #[cfg(feature = "hyper")]
-            IncomingBodyInner::Tcp(body) => Pin::new(body).poll_frame(cx).map_err(Into::into),
-            #[cfg(feature = "h3")]
-            IncomingBodyInner::Quic(body) => Pin::new(body).poll_frame(cx).map_err(Into::into),
-            IncomingBodyInner::Empty => Poll::Ready(None),
-        }
-    }
+    type Error = IncomingBodyError;
 
     fn is_end_stream(&self) -> bool {
-        match &self.inner {
-            #[cfg(feature = "hyper")]
-            IncomingBodyInner::Tcp(body) => body.is_end_stream(),
-            #[cfg(feature = "h3")]
-            IncomingBodyInner::Quic(body) => body.is_end_stream(),
-            IncomingBodyInner::Empty => true,
+        #[allow(unreachable_patterns)]
+        match self {
+            #[cfg(any(feature = "http1", feature = "http2"))]
+            IncomingBody::Hyper(body) => body.is_end_stream(),
+            #[cfg(feature = "http3")]
+            IncomingBody::Quic(body) => body.is_end_stream(),
+            _ => false,
         }
     }
 
-    fn size_hint(&self) -> http_body::SizeHint {
-        match &self.inner {
-            #[cfg(feature = "hyper")]
-            IncomingBodyInner::Tcp(body) => body.size_hint(),
-            #[cfg(feature = "h3")]
-            IncomingBodyInner::Quic(body) => body.size_hint(),
-            IncomingBodyInner::Empty => http_body::SizeHint::with_exact(0),
-        }
-    }
-}
-
-pin_project_lite::pin_project! {
```

Comment:
I think we should keep this somehow it's useful to track how many bytes were sent down the wire.